src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,215 @@
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            Node<T> node = new Node<>(null,value,null);
+            head = node;
+            tail = node;
+            size++;
+        } else {
+            Node<T> node = new Node<>(tail,value,null);
+            tail.next = node;
+            tail = node;
+            size++;
+        }
+
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> node;
+
+        if (index == 0) {
+            node = new Node<>(null,value,head);
+
+            if (head != null) {
+                node.next = head;
+                head.prev = node;
+            } else {
+                head = node;
+                tail = node;
+            }
+            head = node;
+            size++;
+        } else {
+            if (index == size) {
+                node = new Node<>(tail,value,null);
+                node.prev.next = node;
+                tail = node;
+                size++;
+            } else if (index > size || index < 0) {
+                throw new IndexOutOfBoundsException();
+            } else {
+                node = head;
+                for (int i = 0; i < size; i++) {
+                    if (i == index) {
+                        node = new Node<>(node.prev,value,node);
+                        node.prev.next = node;
+                        break;
+                    }
+                    node = node.next;
+                }
+                size++;
+            }
+
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
+        Node<T> node = head;
+        if (index > 0 || index > (size - 1)) {
+            for (int i = 0; i < size; i++) {
+                if (i == index) {
+                    return node.value;
+                } else {
+                    if (node.equals(tail)) {
+                        throw new IndexOutOfBoundsException("size; " + size + "index: " + index);
+                    }
+                    node = node.next;
+                }
+            }
+        } else if (index == 0) {
+            return head.value;
+        } else {
+            throw new IndexOutOfBoundsException("size; " + size + "index: " + index);
+        }
         return null;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = head;
+        T oldValue = null;
+        if ((size - 1) > index) {
+            for (int i = 0; i < size; i++) {
+                if (i == index) {
+                    oldValue = node.value;
+                    node.value = value;
+                    break;
+                }
+                node = node.next;
+            }
+            return oldValue;
+        } else {
+            throw new IndexOutOfBoundsException("size; " + size + "index: " + index);
+        }
+
     }
```

Comment:
Similar to the `get(int index)` method, the `set(T value, int index)` method could also benefit from iterating only over the first or second half of the list, depending on the index value.